src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1285,21 +1285,149 @@ internal bool HasCollectionExpressionApplicableAddMethod(SyntaxNode syntax, Type
+            if (memberAccess is not BoundMethodGroup methodGroup)
+            {
+                return false;
+            }
+
+            Debug.Assert(methodGroup.SearchExtensionMethods);
+
+            if (methodGroup.ResultKind == LookupResultKind.Viable &&
+                methodGroup.Methods.Any(m => isApplicableAddMethod(m.GetConstructedLeastOverriddenMethod(accessingTypeOpt: null, requireSameReturnType: false), expectingExtensionMethod: false)))
+            {
+                return true;
+            }
+
+            if (methodGroup.SearchExtensionMethods &&
+                hasApplicableAddExtensionMethod(syntax, implicitReceiver))
+            {
+                return true;
+            }
+
+            return false;
+
+            bool hasApplicableAddExtensionMethod(SyntaxNode syntax, BoundObjectOrCollectionValuePlaceholder implicitReceiver)
+            {
+                foreach (var scope in new ExtensionMethodScopes(this))
+                {
+                    var methodGroup = MethodGroup.GetInstance();
+                    PopulateExtensionMethodsFromSingleBinder(scope, methodGroup, syntax, implicitReceiver, rightName: methodName, typeArgumentsWithAnnotations: default, BindingDiagnosticBag.Discarded);
+                    bool anyApplicable = methodGroup.Methods.Any(m => isApplicableAddMethod(m, expectingExtensionMethod: true));
+                    methodGroup.Free();
+                    if (anyApplicable)
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            }
+
+            bool isApplicableAddMethod(MethodSymbol method, bool expectingExtensionMethod)
+            {
+                if (method.IsStatic != expectingExtensionMethod)
+                {
+                    return false;
+                }
+
+                var parameters = method.Parameters;
+                int valueIndex = expectingExtensionMethod ? 1 : 0;
+                int requiredLength = valueIndex + 1;
+                if (parameters.Length < requiredLength)
+                {
+                    return false;
+                }
+
+                // Any trailing parameters must be optional or params.
+                int optionalLength = parameters.Length - (OverloadResolution.IsValidParams(this, method) ? 1 : 0);
+                for (int i = requiredLength; i < optionalLength; i++)
+                {
+                    if (parameters[i] is not ({ IsOptional: true, RefKind: RefKind.None or RefKind.In or RefKind.RefReadOnlyParameter }))
+                    {
+                        return false;
+                    }
+                }
+
+                // Extension method 'this' parameter must be by value, ref, in, or ref readonly.
+                if (expectingExtensionMethod)
+                {
+                    if (parameters[0].RefKind is not (RefKind.None or RefKind.Ref or RefKind.In or RefKind.RefReadOnlyParameter))
```

Comment:
>RefKind.Ref or RefKind.In or RefKind.RefReadOnlyParameter [](http://example.com/codeflow?start=70&length=57)

Are we testing extension methods with these modifiers, including when parameter type is not a struct?

Reply:
`AddMethod_Extension_08A`, `_08B`.